atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -224,35 +236,63 @@ public List<OpenTransaction> startTransactions(List<? extends PreCommitCondition
+
+                                transaction.onCommitOrAbort(transaction::reportExpectationsCollectedData);
+                                transaction.onCommitOrAbort(condition::cleanup);
+                                transaction.onCommitOrAbort(
+                                        () -> lockWatchManager.requestTransactionStateRemovalFromCache(
+                                                response.startTimestampAndPartition()
+                                                        .timestamp()));
+                                // N.B. register `onTransactionCommit` after `requestTransactionStateRemovalFromCache`
+                                // `onCommitOrAbort` is FILO, and we need to run `onTransactionCommit` before
+                                // `requestTransactionStateRemovalFromCache`
+                                transaction.onCommitOrAbort(() -> {
+                                    if (transaction.isDefinitivelyCommitted()) {
+                                        lockWatchManager.onTransactionCommit(transaction.getTimestamp());
+                                    }
+                                });
                                 return new OpenTransactionImpl(transaction, immutableTsLock);
                             })
                     .collect(Collectors.toList());
             openTransactionCounter.inc(transactions.size());
             return transactions;
         } catch (Throwable t) {
-            responses.forEach(response -> lockWatchManager.requestTransactionStateRemovalFromCache(
-                    response.startTimestampAndPartition().timestamp()));
-            timelockService.tryUnlock(responses.stream()
+            // In case of failure, we need to remove the transaction from our local lock-watch cache and unlocking
+            // immutable timestamp.
+            // Note that in case we don't throw, the immutable timestamp is registered to be unlocked in the
+            // SnapshotTransaction constructor. But in case of failure, we need to manually unlock it here.
+
+            Closer closer = Closer.create(); // N.B. using closer to run all cleanup tasks even if one cleanup throws
+            responses.forEach(
+                    response -> closer.register(() -> lockWatchManager.requestTransactionStateRemovalFromCache(
+                            response.startTimestampAndPartition().timestamp())));
+            closer.register(() -> timelockService.tryUnlock(responses.stream()
                     .map(response -> response.immutableTimestamp().getLock())
-                    .collect(Collectors.toSet()));
+                    .collect(Collectors.toSet())));
+
+            try {
+                closer.close();
+            } catch (IOException e) {
+                t.addSuppressed(e);
+                log.info("Failed to cleanup startTransaction resources on startTransaction failure", t);
```

Comment:
Yeah, you're doing this again which makes me think my original comments are valuable.